### ISSUE TITLE
Annotations for Nullable and NonNull return values / parameters / fields

### ISSUE THREAD

**Issue by MrStahlfelge** [libgdx/libgdx#6195]
For some time, we have a `@Null` ([source](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/utils/Null.java)) annotation in libGDX. This triggered a discussion on the Discord server recently, because some of us think that the existing annotation could be improved. Our pain points here are the following:

* There is no NonNull annotation guaranteeing non-null values
* The Null annotation doesn't match the general convention that it is named "Nullable"

Moreover, we think that there are some more benefits that could be achieved by such an annotation. Nullable and NonNull annotations can be used by the major IDEs and by Kotlin to determine nullability. The Kotlin compiler even infers the types needed for params and return values in Java interoperability from these annotations, which - in case of `NonNull` makes it possible to reduce a lot of Null-Checks.

As far as we could find out, our annotations have to match the following to be compatible to the following products:

* Eclipse: Nullable/NonNull annotations are fully configurable
* Kotlin: the annotations have to be in certain packages to be used. The [compiler source](https://github.com/JetBrains/kotlin/blob/master/core/compiler.common.jvm/src/org/jetbrains/kotlin/load/java/JvmAnnotationNames.kt) tells us which can be used. It depends on certain package names: `androidx.annotation.Nullable, javax.annotation.Nullable, edu.umd.cs.findbugs.annotations.Nullable, io.reactivex.annotations.Nullable` and more
* IntelliJ/Android Studio: Not 100% sure, but it seems that it accepts the same annotations as Kotlin does

So far so good. If we decide to match the world outside libGDX, there are some more points to decide:
* Which annotation (package) should be used?
* Should we add a dependency, or should we add the annotations ourselves?

@obigu found a great [StackOverflow post on the first topic](https://stackoverflow.com/a/42695253/7487013), listing from which standpoints the decision can be made and how the candidates perform.

Personally, I think we should not adopt annotations retained at runtime - most games won't need this. It looks like the jsr305 annotations are still the best to use, but we might missed some needs so this is intended to start a discussion.

**Comment by PokeMMO** [comment]
>Should we add a dependency, or should we add the annotations ourselves?

Imo, we should not be shipping a class in any of those packages. If we want to match those packages, we need to add them as a dependency.

**Comment by obigu** [comment]
Even if a 3rd party more widely supported annotation would add some benefits (such as by default static code analisys), unfortunately, IMO, they are not good enough to "pollute" framework dependencies plus it adds several potential issues/risks listed on the SO link comments from users on `javax.annotation`).

One think that I do believe there's consensus about is to rename the current `Null` annotation to `Nullable` as it makes more much more sense semantically.

**Comment by cypherdare** [comment]
I experimented with adding Jetbrains annotations in my Java libraries, and I found that adding both Nullable and NonNull to be onerous. I decided it would only be worthwhile to add @Nullable for return values and @NonNull for parameters. This makes it risk-free because the worst the user can do is use unnecessary null-checks. 

My point is, I don't think it is worthwhile if we have to annotate everything. If we really want full null checking, one of the solutions like JSR-305 would be much preferable because it lets you set package defaults. 

If we're mostly thinking of Kotlin here, we might consider simply waiting for their feature for making your own annotations. 

Ultimately I removed the annotations entirely because they are difficult to get working with GWT. I personally have found that the null safety of API calls has limited usefulness. In my experience, NPE problems have happened because of slip-ups in my project code, but when interacting with library functions I always assume non-null is required of parameters unless the docs specify using null to turn something off. libGDX returns null almost nowhere.

**Comment by NathanSweet** [comment]
### May be null

> I always assume non-null is required of parameters unless the docs specify using null to turn something off.

IMO, this is how I think it should work. All method parameters and return values should be assumed to be non-null unless 1) annotated with `@Null`/`@Nullable`, or 2) specified that they can be null in the javadocs. In both cases the coder needs to do some work and I find applying an annotation is easier.

Note javadoc is still needed when it is not obvious what the null means. Often it's "not found" and no javadocs are needed, but sometimes there are other reasons that do need javadocs, a parameter can only be null in certain cases, etc.

### Naming

I prefer `@Null` because it is short. It's pretty obvious that it doesn't mean the value is _always_ null. I don't think naming it `@Nullable` is any more descriptive, nor do I find it important that most other projects prefer `@Nullable`. Since there is no standard, tools should be configurable.

### Non-null

I am very much against using a `@NotNull` annotation. As mentioned above, everything should already be assumed non-null. The common case does not need documentation. The possibility of being null is the special case.

IMO the effort to add it almost everywhere and the clutter doing that adds is unreasonable.  Being forced to write it explicitly is a time waste. Rather than doing busy work and cluttering up our code, tools should assume everything is non-null.

### IDEs

The reason I use a nullable annotation is because it is easier than javadocs. IDE assistance provided by a nullable annotation is marginal at best. If left as a warning it occurs way too often and trying to satisfy all occurrences of those warnings would be a waste of time. Also the IDE gets it wrong sometimes, so to satisfy the warning you'd need to jump so some ridiculous hoops (which will change as the compilers do) or add a suppress annotation. IMO none of that is worthwhile. In Eclipse I change it to `info` level and then I exclude `info` from the `Problems` view. This means I get underlines in the source code, which do help a little bit to highlight things that may be null, but otherwise I am not inconvenienced.

**Comment by PokeMMO** [comment]
>As mentioned above, everything should already be assumed non-null.

I disagree, using this assumption on a library you do not explicitly read will lead to many unnecessary NPEs. I would think most programmers are used to assuming null for safety, and only assuming non-null when it is verified or explicitly guarnteed to them (In the form of an annotation.)

When properly setup, a Nonnull annotation is more useful to me than a Nullable, as it will explicitly tell me "X cannot be null here" and to suggest removing a redundant null check. With no annotation, the IDE will not be able to make this suggestion.

>Also the IDE gets it wrong sometimes, so to satisfy the warning you'd need to jump through some ridiculous hoops
You're correct on this, but this only applies to Nullable. Nonnull will not make you do extra work.

**Comment by NathanSweet** [comment]
> I disagree, using this assumption on a library you do not explicitly read will lead to many unnecessary NPEs. 

I didn't suggest using a library without reading its documentation. The non-null assumption is for libgdx code, my own code, and anywhere else it holds true. I contend that it's a good convention everyone should follow and tools should support.

> When properly setup, a Nonnull annotation is more useful to me than a Nullable, as it will explicitly tell me "X cannot be null here" and to suggest removing a redundant null check. With no annotation, the IDE will not be able to make this suggestion.

Is removing redundant null checks terribly useful? It comes at the cost of pasting `@NonNull` almost everywhere.

Knowing something can be null is much more important, since it is part of method contracts and can avoid crashes. Though as mentioned, I would not use it solely for IDE assistance (which is poor), rather I use it for easier documentation, which must be done via javadocs if not using annotations for it.

> Nonnull will not make you do extra work.

I agree the IDE won't cause you to jump through hoops to satisfy non-null annotations like it does nullable. The extra work is still there though, in the form of 1) pasting it _everywhere_ and updating contributions to have it, and 2) reading past it millions of times, since code is read a lot more than it is written.

**Comment by obigu** [comment]
> ### Naming
> I prefer `@Null` because it is short. It's pretty obvious that it doesn't mean the value is _always_ null. I don't think naming it `@Nullable` is any more descriptive, nor do I find it important that most other projects prefer `@Nullable`. Since there is no standard, tools should be configurable.

I think it's worth listing the pros and cons the cons of each option.

**Pros of @Null**
- 4 characters long

**Cons of @Null**
- Breaks convention
- Shares name with `javax.validation.constraints.Null` annotation (which actually checks for null value)
- Semmantically wrong

**Pros of @Nullable**
- It's the naming by widely adopted convention (even if it's not a standard as JSR-305 has not been adopted)
- Self-documenting code

**Cons of @Nullable**
- 8 characters long

IMO the fact that it's shorter is much less important compared to the other points.

**Comment by NathanSweet** [comment]
I would say both are self documenting.

> 8 characters long

The shorter name is nice for something repeated so many times throughout most code. libgdx currently has 452 `@Null` annotations, even though so far it's only been applied to scene2d and (some of?) utils.

> javax.validation.constraints.Null

This seems to be an EE class that I've never seen on my classpath.

> Semantically wrong

Sort of. It's fine if considered a simple tag for parameters, return types, and fields. It makes you think about nulls is pretty obvious what it means. The "code reads like a sentence" stuff never works out.

> Breaks convention

True. The next question is, what are the pros/cons of following this convention? One consideration is IDE support, though I don't think we should weight decisions much based on this given tools are expected to be more flexible in the future and IDE assistance for this feature is secondary. Other than I'm not sure there are any advantages to the convention.

Of the pros/cons, for me it comes down to semantically wrong vs length. I personally find it easy to get over the semantics. I'm not super excited about the longer name being sprinkled everywhere, though it is somewhat mitigated that the typical method declaration only having a single `@Null` annotation. So far there are only a few places with more than two. For those the comparison is eg:
```
public ImageButtonStyle (@Null Drawable up, @Null Drawable down, @Null Drawable checked, @Null Drawable imageUp,
	@Null Drawable imageDown, @Null Drawable imageChecked) {
```
And:
```
public ImageButtonStyle (@Nullable Drawable up, @Nullable Drawable down, @Nullable Drawable checked, @Nullable Drawable imageUp,
	@Nullable Drawable imageDown, @Nullable Drawable imageChecked) {
```
Both are pretty nasty TBH, with `@Null` being a tiny bit less. This kind of parameter list is better avoided entirely. I'm not sure what other places in libgdx it will come up.

Ultimately if I'm the only one who prefers `@Null` then I'm willing to concede and we can name it `@Nullable`. However, I feel much more strongly against using `@NotNull`/`@NonNull`.

**Comment by crykn** [comment]
Just to give another opinion on this: Personally I’m using the jsr305 annotations in all of my projects. Same as Nate I’m only using them for documentation purposes; I like being able to quickly see whether null is an expected value for a certain variable/parameter.

As the others, I’m in favour of using `@Nullable`, (a) because of it being convention, which makes it easier for newcomers to acquaint themselves with the libgdx code and (b) because it is the only way to give Kotlin users compiler support. The latter also means that we have to use a Nullable annotation in a certain package and as PokeMMO pointed out that means we should most certainly use the corresponding dependency.

Regarding `@NonNull`: I have to agree with Nate on this one. Null not being a valid value should be the default assumption. I think it’s tedious to add `@NonNull` to every variable and it would make the code _a lot_ more cluttered than just using `@Nullable`.

**Comment by MrStahlfelge** [comment]
You all assume that return values of called methods are non-null? I think that's a very bold assumption, I always assume that returned objects may be null and hence think NonNull annotation on methods is very valuable.

**Comment by cypherdare** [comment]
You can't do that with _any_ library, but I know it's ok with libGDX because of how it's designed. I do briefly skim the docs every time I call a method I'm not familiar with to see if it says "May be null." But my assumption is if a method's docs don't say "May be null," it would be a breaking change for it to someday start returning null, and that would need to be documented in the change log as a breaking change. 

I vote for `@Nullable`. I'd even vote for `@$` or `@N` over `@Null` if we're really after conciseness. `@Null` is just so semantically wrong that it bugs me each time I see it.

**Comment by obigu** [comment]
> I would say both are self documenting.

Self documenting, imo, is all about having code that self describes and following conventions so that users get what it does without having to read the docs. I'd say a percentage of users that see @Null go to the annotation source code to check what it means (I did).

> > javax.validation.constraints.Null
> 
> This seems to be an EE class that I've never seen on my classpath.

True, the `javax.validation.constraints` package is widely used on Java backend apps (the kind of apps that use Spring, Hibernate and stuff). For the ones of us who come from a Java backend background this package is not unfamiliar.

> 
> > Semantically wrong
> 
> Sort of. It's fine if considered a simple tag for parameters, return types, and fields. It makes you think about nulls is pretty obvious what it means. The "code reads like a sentence" stuff never works out.

I agree it could be worse and be named `@Nl` for instance, but, it is semantically wrong, the word "null" doesn't convey the behaviour the annotation describes but "nullable" does.

> 
> Ultimately if I'm the only one who prefers `@Null` then I'm willing to concede and we can name it `@Nullable`. However, I feel much more strongly against using `@NotNull`/`@NonNull`.

Thanks for being open to change.

**Comment by NathanSweet** [comment]
Alright, unless a bunch of lurkers that prefer `@Null` speak up soon, I guess we'll rename it to `@Nullable`.

**Comment by noblemaster** [comment]
Either `@Nullable` or `@Null` is fine with me. Although `@Nullable` seems to be more the standard. Also, it probably even makes sense to consider **cypherdare**'s suggestion using `@$` or `@N` as it would prevent method parameters not to become overly long.
 
I am against the `@NonNull`/`NotNull` as it just adds clutter. Everything that doesn't have a `@Nullable` annotation can just be considered never to not return a `null` value. We also don't have a `@NotAnOverride` annotation for methods. It's just assumed a method isn't overwritten if there isn't a `@Override` annotation.

**Comment by MobiDevelop** [comment]
Unless it's libgdx, then it might or might not be an override whether the annotation I there or not.

**Comment by cypherdare** [comment]
Yeah, please don't actually do that. I was just making a point that those are more concise than Null and no less descriptive.

**Comment by tommyettinger** [comment]
If we use JSR305, there's some mechanisms it has to mark all of some "thing" in a package as being `@Nonnull` by default, unless otherwise annotated. This is used in LWJGL3 and we could copy their implementation of https://github.com/LWJGL/lwjgl3/blob/master/modules/lwjgl/core/src/main/java/org/lwjgl/system/NonnullDefault.java . In that case, "thing" means every field, method return, and parameter. This reduced the bloat in my experimental GDX-based collections library quite a bit; [see the changes in this commit](https://github.com/tommyettinger/jdkgdxds/commit/92e1b924ad54896d0741718f2d804dc1f275b456#diff-59f7d1921a50eec5073a9acd965cc2bc), ignoring the HTML docs committed at the top.  I changed the retention policy to `CLASS` instead of `RUNTIME` so there should be no overhead, or nearly none. Of course, a value should only be annotated as `@Nonnull` if the value is either known to never be null or is known that code will malfunction if the value is null. If the opposite assumption is true and we want things to be allowed to be nullable by default, the way LWJGL3 uses will also work, but it's kinda pointless because Java already assumes references are nullable.

**Comment by tommyettinger** [comment]
> It's just assumed a method isn't overwritten if there isn't a @Override annotation.
Uh... you're kidding.

**Comment by crykn** [comment]
To summarise the above: The state of discussion seems to have been that `@Nullable` is preferred over `@Null`, and `@Nonnull` should not be added. This leaves the question: what `@Nullable`annotation are we going to use?

Just creating our own `@Nullable` annotation won’t be supported by the [Kotlin compiler](https://github.com/JetBrains/kotlin/blob/master/core/compiler.common.jvm/src/org/jetbrains/kotlin/load/java/JvmAnnotationNames.kt). And looking at the thumbs ups on @PokeMMO's [comment above](https://github.com/libgdx/libgdx/issues/6195#issuecomment-698915366), I think the majority is in favour of using a dependency instead of just copying a class in a package that is not 'ours'. This leaves us with two sensible options for `@Nullable`:

#### 1) `javax.annotation.Nullable` supplied by `com.google.code.findbugs:jsr305`
- (+) Kotlin compiler & IDE support
- (+) A concise package name
- (+) Can be used to set the non-null-ness for a whole package (via `@TypeQualifierDefault`)
- (~) Is used by LWJGL3, whatever that is worth 
- (~) JSR-305 is abandoned
- (-) There is a split package issue with JDK 9+ [if JAX-WS](https://github.com/google/guava/issues/2960#issuecomment-546713529) is used. However, I don’t think there are a lot of people using JAX-WS and libGDX at the same time and there seems to be a [workaround](https://github.com/google/guava/issues/2960#issuecomment-522446788) involving patching.
- (-) There may be some license problems with [Oracle JDKs](https://stackoverflow.com/questions/13414474/alternative-to-findbugs-defaultannotation-for-javax-annotation-for-fields-and-me/36198568#36198568)

#### 2)  `org.checkerframework.checker.nullness.qual.Nullable` supplied by `org.checkerframework:checker-qual`
- (+) Kotlin compiler & IDE support
- (+) Adopted by a lot of libraries in place of JSR-305 for example by Google’s [Guava](https://github.com/google/guava/issues/2960#issuecomment-350721044)
- (+) Can be used to set the non-null-ness for a whole package (via `@DefaultQualifier`); but this is supposedly [not picked up](https://dzone.com/articles/when-to-use-jsr-305-for-nullability-in-java) by IntelliJ or the Kotlin compiler
- (~) The use of Checker Framework is somewhat recommended by the [Oracle Java Tutorials](https://docs.oracle.com/javase/tutorial/java/annotations/type_annotations.html)
- (-) Has a longer package name

Iirc @Tommyettinger did some testing regarding the size of the dependencies.

<br/>

In the end, I think it doesn’t really matter what annotation we are using - there will always be advantages and disadvantages. We should just chose one and then make the best out of it.
